src/proto/generation.proto

```diff
@@ -240,6 +241,92 @@ message ClassifierParameters {
   optional Action realized_action = 3;
 }
 
+//
+// Transforms
+//
+
+enum BorderMode {
+  BORDER_REFLECT = 0;   // reflect image values across the border
+  BORDER_REPLICATE = 1; // replicate border values outside the image
+  BORDER_WRAP = 2;      // wrap around / tile the image values
+  BORDER_ZERO = 3;      // use 0 for locations outside the image
+}
+
+enum ColorMatchMode {
+  COLOR_MATCH_HSV = 0;  // match hue, saturation, and value histograms
+  COLOR_MATCH_LAB = 1;  // match lightness, a, and b histograms
+  COLOR_MATCH_RGB = 2;  // match red, green, and blue histograms
+}
+
+message TransformAddNoise {
+  float amount = 1; // amount of noise to add
+  uint32 seed = 2;  // random seed for noise generation
+}
+
+message TransformBlend {
+  float amount = 1;     // blend amount from 0 to 1
+  Artifact target = 2;  // target image
+}
+
+message TransformColorMatch {
+  ColorMatchMode color_mode = 1;
+  Artifact image = 2;
+}
+
+message TransformContrast {
+  optional float brightness = 1;  // scale on pixel intensities, 1.0 is no change
+  optional float contrast = 2;    // contrast adjustment, 1.0 is no change
+}
+
+message TransformDepthCalc {
+  optional bool export = 1;        // return the depth map as an artifact
+  optional float blend_weight = 2; // blend factor between AdaBins (0.0) and MiDaS (1.0)
+}
+
+message TransformWarp2d {
+  BorderMode border_mode = 1; // extrapolation of border pixels
+  float translate_x = 4;      // horizontal translation in pixels
+  float translate_y = 5;      // vertical translation in pixels
+  float rotate = 2;           // rotation in degrees
+  float scale = 3;            // scale factor
+}
+
+message TransformWarp3d {
+  BorderMode border_mode = 1; // extrapolation of border pixels
+  float translate_x = 3;      // horizontal translation
```

Comment:
Why are we doing this as individual variables instead of a x, y, z structure? We should not be doing it this way.

Reply:
This has not been addressed.

Reply:
i think the only reason is because that's the convention that developed in the notebook tooling. vectors were the convention for 3d transforms when quaternions were used to specify rotations, but I think it's always been indpendent components for 2D transforms since chia released her notebook. disco's transition to euler angles made it feasible for people to reason about 3D rotations the same way they did about 2D, so here we are.

History aside, one reason to break stuff up like this could be to reduce data transfer, i.e. if only one component is nonzero. but these values are non-optional in the current spec, so that's moot.

I'm on board with changing this to an array. it won't be in line with how users are currently constructing these parameters or how we're consuming them in the generator, but that's not a good reason not to iterate towards a better spec.